### PR TITLE
Create request controllers

### LINKED ISSUE #8: BE: Request Controllers
We need an endpoint for the following related to requests: 
\- GET : list of requests for logged in user
\- POST : Create a new request
\- UPDATE  : Update request with approved or decline

### PR DESCRIPTION
**Pull request by denjj**
### What this PR does (required):
- Creates and integrates api endpoint for requests
- Closes #8 

### Any information needed to test this feature (required):
- Tested with request model #53 
- With an api testing tool, login as an existing user to retrieve an authorization token
- Perform POST /requests/create with **sitterId**, **start**, and **end** in the body to create a new request
Start and end parameters are Javascript dates, a Number that represents milliseconds since 1 January 1970 UTC.
- Perform GET /requests to load requests for the logged in user
- Perform PUT /requests/approval with **requestId**, and **accept** in the body to accept or deny a request
Accept is a boolean, where true will accept the request and false will deny it

### CODE REVIEW

File: server/controllers/request.js [hunk 1]
```diff
@@ -0,0 +1,105 @@
+const Request = require("../models/Request");
+const asyncHandler = require("express-async-handler");
+
+// @route GET /requests
+// @desc load requests for logged in user
+// @access Private
```
Comment (ezeilo-su) [inline server/controllers/request.js:6]: Good documentation

File: server/routes/request.js [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+  loadRequests,
+  createRequest,
+  approveRequest,
+} = require("../controllers/request");
+
+router.route("/").get(protect, loadRequests);
+
+router.route("/create").post(protect, createRequest);
+
+router.route("/approval").put(protect, approveRequest);
```
Comment (ezeilo-su) [inline server/routes/request.js:14]: :+1:

**Review by ezeilo-su — COMMENTED** [review]
Good work with this one!

File: server/controllers/request.js [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+exports.loadRequests = asyncHandler(async (req, res, next) => {
+  const requests = await Request.find({ userId: req.user.id });
+
+  if (!requests) {
```
Comment (rajivtitus) [inline server/controllers/request.js:10]: The mongoose `find()` query returns an empty list if no profiles are found, to remain consistent with the response format it is recommended we return the empty list to the FE (where we can conditionally render a message if no profiles are found)

File: server/controllers/request.js [hunk 4]
```diff
@@ -0,0 +1,105 @@
+const Request = require("../models/Request");
+const asyncHandler = require("express-async-handler");
+
+// @route GET /requests
+// @desc load requests for logged in user
+// @access Private
+exports.loadRequests = asyncHandler(async (req, res, next) => {
+  const requests = await Request.find({ userId: req.user.id });
```
Comment (rajivtitus) [inline server/controllers/request.js:8]:
```suggestion
  const requests = await Request.find({ sitterId: req.user.id });
```
If we wish to return request for a sitter, we'll need to filter by `sitterId`

File: server/controllers/request.js [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+
+  if (startDate.getTime() >= endDate.getTime()) {
+    res.status(400);
+    throw new Error("End date must be later than start date");
+  }
+
+  const sitter = await User.findById(sitterId);
+
+  if (!sitter) {
+    res.status(400);
```
Comment (rajivtitus) [inline server/controllers/request.js:50]:
```suggestion
    res.status(404);
```

File: server/controllers/request.js [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+      requests: requests,
+    },
+  });
+});
+
+// @route POST /requests/create
+// @desc create a new request
+// @access Private
+exports.createRequest = asyncHandler(async (req, res, next) => {
+  const { sitterId, start, end } = req.body;
```
Comment (rajivtitus) [inline server/controllers/request.js:26]: Optional: We can also validate the request body params and return a `400` if they're missing (ideally we would want to use a middleware to do so - express validator should be a part of the pkg bundle)

File: server/routes/request.js [hunk 7]
```diff
@@ -0,0 +1,16 @@
...
+const protect = require("../middleware/auth");
+const {
+  loadRequests,
+  createRequest,
+  approveRequest,
+} = require("../controllers/request");
+
+router.route("/").get(protect, loadRequests);
+
+router.route("/create").post(protect, createRequest);
```
Comment (rajivtitus) [inline server/routes/request.js:12]:
```suggestion
router.route("/").post(protect, createRequest);
```
We can simplify these routes as the HTTP verbs indicate the purpose

File: server/routes/request.js (hunk 2 above)
Comment (rajivtitus) [inline server/routes/request.js:14]:
```suggestion
router.route("/:requestId").put(protect, approveRequest);
```
Let's use path params as we're identifying and modifying a single resource :)

File: server/controllers/request.js [hunk 8]
```diff
@@ -0,0 +1,105 @@
...
+    accepted = JSON.parse(accept.toLowerCase());
+    if (typeof accepted !== "boolean") {
+      throw new TypeError();
+    }
+  } catch (err) {
+    res.status(400);
+    throw new Error("Invalid request data");
+  }
+
+  try {
```
Comment (rajivtitus) [inline server/controllers/request.js:89]: We don't a try/catch here as the asyncHandler will handle promise rejections for us

File: server/controllers/request.js [hunk 9]
```diff
@@ -0,0 +1,105 @@
...
+    const request = await Request.findById(requestId);
+
+    request.accepted = accepted;
+    request.declined = !accepted;
+    await request.save();
```
Comment (rajivtitus) [inline server/controllers/request.js:94]: Only the sitter who owns the request should be able to modify it

File: server/controllers/request.js [hunk 10]
```diff
@@ -0,0 +1,105 @@
...
+    res.status(400);
+    throw new Error("Invalid request data");
+  }
+});
+
+// @route PUT /requests/approval
+// @desc update a request with approved or declined
+// @access Private
+exports.approveRequest = asyncHandler(async (req, res, next) => {
+  const { requestId, accept } = req.body;
```
Comment (rajivtitus) [inline server/controllers/request.js:77]: We would also need the functionality to decline a request :)

If we use a single field like `status` to keep track of the status - it should make things easier for us (please see comment in request model PR)

**Review by rajivtitus — COMMENTED** [review]
Excellent work 💯 
Just a few comments for you to have a look and address, please let me know if you have questions.

File: server/controllers/request.js [hunk 11]
```diff
@@ -0,0 +1,92 @@
...
+  const { requestId } = req.params;
+
+  const request = await Request.findById(requestId);
+
+  if (!request) {
+    res.status(404);
+    throw new Error("Request not found");
+  }
+
+  if (request.sitterId.toString() !== req.user.id) {
```
Comment (rajivtitus) [inline server/controllers/request.js:77]: 💯

**Review by rajivtitus — APPROVED** [review]
Excellent work with the changes 💯

**Pull request merged**